### PR TITLE
Add DBSync Module

### PR DESCRIPTION
**Pull request by codchen**
## Describe your changes and provide context
Added a new `internal` module called `dbsync` that enables syncing application DB files directly. Its discovery mechanism resembles that of `statesync` with some simplification. Still a WIP

## Testing performed to validate your change
unit tests
tested in loadtest cluster

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -1257,6 +1260,43 @@ func (cfg *InstrumentationConfig) ValidateBasic() error {
 	return nil
 }
 
+type DBSyncConfig struct {
+	Enable               bool          `mapstructure:"db-sync-enable"`
```
Comment (yzang2019) [inline config/config.go:1264]: Nice! Thanks for the changing the config name

File: abci/types/application.go [hunk 2]
```diff
@@ -31,6 +32,7 @@ type Application interface {
 	OfferSnapshot(context.Context, *RequestOfferSnapshot) (*ResponseOfferSnapshot, error)                // Offer a snapshot to the application
 	LoadSnapshotChunk(context.Context, *RequestLoadSnapshotChunk) (*ResponseLoadSnapshotChunk, error)    // Load a snapshot chunk
 	ApplySnapshotChunk(context.Context, *RequestApplySnapshotChunk) (*ResponseApplySnapshotChunk, error) // Apply a shapshot chunk
+	LoadLatest(context.Context, *RequestLoadLatest) (*ResponseLoadLatest, error)
```
Comment (yzang2019) [inline abci/types/application.go:35]: Can we add a small comment for what this interface do? Like the above ones?

File: config/config.go [hunk 3]
```diff
@@ -1257,6 +1260,43 @@ func (cfg *InstrumentationConfig) ValidateBasic() error {
...
+}
+
+func DefaultDBSyncConfig() *DBSyncConfig {
+	return &DBSyncConfig{
+		Enable:               false,
+		SnapshotInterval:     0,
+		SnapshotDirectory:    "",
+		TimeoutInSeconds:     0,
+		NoFileSleepInSeconds: 5,
+		FileWorkerCount:      0,
```
Comment (yzang2019) [inline config/config.go:1283]: Can we have better and more common defaults here? I'd suggest let's provide some meaning ful default value for:

- SnapshotInterval
- TimeoutInSeconds
- NoFileSleepInSeconds
- FileWorkerCount
- FileWorkerTimeout
- TrustPeriod

File: internal/dbsync/reactor.go [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+		chainID:       chainID,
+		providers:     make(map[types.NodeID]*light.BlockProvider),
+		eventBus:      eventBus,
+		config:        config,
+		postSyncHook:  postSyncHook,
+	}
+	syncer := NewSyncer(logger, config, baseConfig, reactor.requestMetadata, reactor.requestFile, reactor.commitState, reactor.postSync)
+	reactor.syncer = syncer
+
+	reactor.BaseService = *service.NewBaseService(logger, "BlockSync", reactor)
```
Comment (yzang2019) [inline internal/dbsync/reactor.go:159]: This should be DbSync?

**Review by yzang2019 — COMMENTED** [review]
Overall looks great! No blocking comments, left some minor things though

File: abci/types/application.go (hunk 2 above)
Comment (codchen) [reply abci/types/application.go:35]: added

File: config/config.go (hunk 3 above)
Comment (codchen) [reply config/config.go:1283]: updated

File: internal/dbsync/reactor.go (hunk 4 above)
Comment (codchen) [reply internal/dbsync/reactor.go:159]: nice catch

**Review by yzang2019 — APPROVED** [review]

**Pull request merged**